skabase/SKALogger/SKALogger/SKALogger.py

```diff
@@ -101,60 +133,171 @@ def delete_device(self):
 
     @command(
     dtype_in=('str',), 
-    doc_in="none", 
+    doc_in="Details of timestamp, logging level, source device and message.", 
     )
     @DebugIt()
     def Log(self, argin):
         # PROTECTED REGION ID(SKALogger.Log) ENABLED START #
-        ###Log (timestamp,level,source,message)
-        source_device =  argin[2]
-        message = argin[3]
+        """
+        A method of LogConsumer Interface, to enable log viewer.
+        """
+        logLevel = argin[1]
+        logSource = argin[2]
+        logMessage = argin[3]
+        tangoLogLevel = {"FATAL": 1, "ERROR": 2, "WARN": 3, "INFO": 4, "DEBUG": 5}
+        levelNumber = tangoLogLevel[logLevel]
+        device = DeviceProxy(logSource)
```

Comment:
> It doesn't seem like a good idea to instantiate a TANGO client back to the source and read an attribute (lines 153/156) for every single log message. This is likely to be a bottleneck when we have a high volume of log messages.

>The Logger device should maintain the log levels - they should only be modified via this Logger device's Set*LoggingLevel commands. So the Logger device is the "owner" of the log levels.

As a LogConsumer device, the file must contain Log function. For this, it has been implemented in SKALogger base class. Since the source devices are separate, we had to create proxy and set levels from there. What I understand from your suggestion, we may need to implement a source dictionary that contains source names and their respective log levels. The Log function will refer this dictionary while logging the message to the log target. If my understanding is correct, I consider this as enhancement and hence we will keep this as backlog item.

> A workaround, using the current TANGO core implementation, would have to bypass the standard TANGO Logging System (TLS). So SKABaseDevice needs to intercept calls to TLS and instead execute Log commands directly to both the central and element logger, depending on the configured levels. In the case of the Logger classes (which inherit from SKABaseDevice), we'd need to disable this interception.

>With such a workaround, it would be easier if the *LoggingLevel attributes stay read-write.

There is a function called dev_log_msg in SKATestDevice currently which logs the message to different logging target based on separate logging levels. I suppose if we mode this function into SKABaseDevice then the workaround you are suggesting can be implemented (I am anyways planning to move it as per our conversation [here](https://github.com/ska-telescope/lmc-base-classes/pull/84#discussion_r243266475) ). Correct me if there is difference in our understanding.

Reply:
Yes, using a local lookup instead of connecting to the client again with DeviceProxy in the `Log` function can be seen as an enhancement, and kept on the backlog.  Very important though.